workload/schemachange: enable inserts

annrpom opened the pull request:
### workload/schemachange: use a pseudo-rand rng

This patch replaces the rng we used for the
schemachange workload with a rand that can
be seeded from the `COCKROACH_RANDOM_SEED'
environment variable and properly printed
out for later use.

Epic: none

Release note: None

----

### workload/schemachange: fix existing table search

With the addition of workers specified in table names
to make our workload more deterministic, this patch
fixes the regex match we had when we look for existing
tables.

Epic: none

Release note: None

----

### workload/schemachange: enable inserts

This patch enables inserts inside of the randomized schema change workload.

Fixes: #91863
Epic: CRDB-19168

Release note: None

annrpom committed:
workload/schemachange: enable inserts

This patch enables inserts inside of the randomized schema change
workload.

Epic: CDRB-19168

Fixes: #91863
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/114052)
<!-- Reviewable:end -->

annrpom commented:
***UNEXPECTED ERROR; Received an unexpected execution error.

most test runs fail with this - sometimes we get a syntax error from the generated expression (computed column) name, but investigating to see if this is another small syntax error or an actual bug

annrpom commented:
workload is fine in the sense that it is not failing, but it is also not generating interesting stmts now; i don't think it's related to inserts since this started happening after i rebased, but let me double-check

annrpom committed:
workload/schemachange: fix existing table search

With the addition of workers specified in table names
to make our workload more deterministic, this patch
fixes the regex match we had when we look for existing
tables.

Epic: none

Release note: None

annrpom commented on pkg/sql/randgen/schema.go:
```diff
@@ -464,7 +466,7 @@ func GenerateRandInterestingTable(db *gosql.DB, dbName, tableName string) error
 // column, with a random type and nullability.
 func randColumnTableDef(rng *rand.Rand, tableIdx int, colSuffix string) *tree.ColumnTableDef {
-	g := randident.NewNameGenerator(&nameGenCfg, rng, fmt.Sprintf("col%d_", tableIdx))
+	g := randident.NewNameGenerator(&nameGenCfg, rng, fmt.Sprintf("col%d", tableIdx))
```
GenerateOne down below will add a `_` to the suffix. This would result in column names like `col3__w0_234`. This just fixes that part

annrpom commented:
We're almost there! `ERROR: cannot write directly to computed column "col1_w0_4" (SQLSTATE 55000)` crops up a fair amount

annrpom commented on pkg/workload/schemachange/error_screening.go:
```diff
@@ -1694,3 +1616,73 @@ func (og *operationGenerator) databaseHasTablesWithPartitioning(
+                          FROM unique_indexes AS idx
+                               INNER JOIN columns AS c ON idx.col_id = c.col_id
+                     )
+      SELECT '(' || array_to_string(array_agg(expr), ', ') || ')' AS final_expr                    
```
This specific change (i.e. we don't use `SELECT ARRAY['(' || array_to_string(array_agg(expr), ', ') || ')'] AS final_expr`) had to happen because casting to an array implicitly improperly escaped the column names - leading us to use queries like `SELECT \"col blah2\" ...` instead of `SELECT "col blah2" ...`

annrpom committed:
workload/schemachange: use a pseudo-rand rng

This patch replaces the rng we used for the
schemachange workload with a rand that can
be seeded from the `COCKROACH_RANDOM_SEED'
environment variable and properly printed
out for later use.

Epic: none

Release note: None

annrpom commented:
Opening for comments - ran the workload locally several times and they have all been passing (i.e. I can't track down the above ^ errors, but I'll keep my eye out).

Monitoring CI

rafiss reviewed (commented):
very exciting! i just had some nits and questions, but it all lgtm

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114052#-:-NqFQ9lC7H7B4dDeNqAA:b-ausxji)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @herkolategan, and @renatolabs)

___
*[`pkg/workload/schemachange/error_screening.go` line 265 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/114052#-NqFQP4l9ZOr6-5VG94v:-NqFQP4l9ZOr6-5VG94w:bmagdnt) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bd4f6357e8079a82d09616fe6437665dc769db6/pkg/workload/schemachange/error_screening.go#L265)):*
> ```Go
> ) (bool, codesWithConditions, error) {
> 	var generatedCodes codesWithConditions
> 	constraints, err := getUniqueConstraintsForTable(ctx, tx, tableName.String())
> ```

nice cleanup :)
___
*[`pkg/workload/schemachange/operation_generator.go` line 3321 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/114052#-NqFOYZO1i7H8Nvxa_VT:-NqFOYZO1i7H8Nvxa_VU:b-j7vf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ead911a3c37e6f13fa89f74faa20accdbf523/pkg/workload/schemachange/operation_generator.go#L3321)):*
> ```Go
>     SELECT table_schema, table_name, column_name, crdb_sql_type, is_nullable
>       FROM information_schema.columns
> 		 WHERE table_name ~ 'table_w[0-9+]_[0-9+]' AND column_name <> 'rowid'
> ```

perhaps we should make it `table_name LIKE 'table%'` instead of `table_name ~ 'table[0-9]+'` -- i noticed that's what the other queries to fetch functions, schemas, types, indexes, etc do. it might be easier to maintain since it's less selective.
___
*[`pkg/workload/schemachange/operation_generator.go` line 3086 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/114052#-NqFPSvc8D-Yqgp6u9Xa:-NqFPSvc8D-Yqgp6u9Xb:b-41a355) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bd4f6357e8079a82d09616fe6437665dc769db6/pkg/workload/schemachange/operation_generator.go#L3086)):*
> ```Go
> 			}
> 			return errors.Mark(
> 				og.WrapWithErrorState(errors.Wrap(err, "***UNEXPECTED ERROR from %s; Received a non pg error."), s),
> ```

nit: what is the `%s` matching?
___
*[`pkg/workload/schemachange/operation_generator.go` line 3208 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/114052#-NqFPb1O6hMpmumjytYi:-NqFPb1O6hMpmumjytYj:brcp088) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bd4f6357e8079a82d09616fe6437665dc769db6/pkg/workload/schemachange/operation_generator.go#L3208)):*
> ```Go
> 	rows, err := tx.Query(ctx, q, tableName)
> 	if err != nil {
> 		return nil, errors.Wrapf(err, "getting table columns from %s", tableName.String())
> ```

do you know why the `.String()` is needed? shouldn't it be calling `String()` anyway because of the `%s` directive?
___
*[`pkg/workload/schemachange/schemachange.go` line 199 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/114052#-NqFMW4o0qUt_7phf6f6:-NqFMW4o0qUt_7phf6f7:b-h4mh4t) ([raw file](https://github.com/cockroachdb/cockroach/blob/84f5446022cc27d2ddb923b850796975d762b61f/pkg/workload/schemachange/schemachange.go#L199)):*
> ```Go
> 	}
> 	stdoutLog := makeAtomicLog(os.Stdout)
> 	rng, seed := randutil.NewPseudoRand()
> ```

oh weird, i thought `NewTestRand` also was seeded by COCKRAOCH_RANDOM_SEED. it looks like ultimately it uses the this global seed, but have you seen otherwise?

https://github.com/cockroachdb/cockroach/blob/397dc0b48ba959278c0ab9e0042bf5b78940b279/pkg/util/randutil/rand.go#L77


<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114052)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @rafiss, and @renatolabs)

___
*[`pkg/workload/schemachange/operation_generator.go` line 3321 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/114052#-NqFOYZO1i7H8Nvxa_VT:-NqTdK9C7G2y5Y1D8aMU:b-xg1l7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ead911a3c37e6f13fa89f74faa20accdbf523/pkg/workload/schemachange/operation_generator.go#L3321)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

perhaps we should make it `table_name LIKE 'table%'` instead of `table_name ~ 'table[0-9]+'` -- i noticed that's what the other queries to fetch functions, schemas, types, indexes, etc do. it might be easier to maintain since it's less selective.
</blockquote></details>

good point, but i think this includes extraneous tables that are auto-generated, correct? like table_columns, table_indexes, table_row_statistics, etc.

i can do `LIKE 'table_w%'` perhaps, assuming that no auto-generated table_w* gets added 
___
*[`pkg/workload/schemachange/operation_generator.go` line 3086 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/114052#-NqFPSvc8D-Yqgp6u9Xa:-NqTZoglEl7oBhalGFbm:b-cf7ruy) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bd4f6357e8079a82d09616fe6437665dc769db6/pkg/workload/schemachange/operation_generator.go#L3086)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: what is the `%s` matching?
</blockquote></details>

done  - thanks for catching! this was from debugging - i wanted to print out the type of op the workload failed
___
*[`pkg/workload/schemachange/operation_generator.go` line 3208 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/114052#-NqFPb1O6hMpmumjytYi:-NqTbNNEETEhTz8BEy5X:b-yjy212) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bd4f6357e8079a82d09616fe6437665dc769db6/pkg/workload/schemachange/operation_generator.go#L3208)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

do you know why the `.String()` is needed? shouldn't it be calling `String()` anyway because of the `%s` directive?
</blockquote></details>

done - also left over from testing
___
*[`pkg/workload/schemachange/schemachange.go` line 199 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/114052#-NqFMW4o0qUt_7phf6f6:-NqTq1S_5pO1UepThMus:b-aqzfx9) ([raw file](https://github.com/cockroachdb/cockroach/blob/84f5446022cc27d2ddb923b850796975d762b61f/pkg/workload/schemachange/schemachange.go#L199)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

oh weird, i thought `NewTestRand` also was seeded by COCKRAOCH_RANDOM_SEED. it looks like ultimately it uses the this global seed, but have you seen otherwise?

https://github.com/cockroachdb/cockroach/blob/397dc0b48ba959278c0ab9e0042bf5b78940b279/pkg/util/randutil/rand.go#L77
</blockquote></details>

Yeah, it does look like NewTestRang gets seeded by `COCKRAOCH_RANDOM_SEED` and then "re-seeded" (?) here https://github.com/cockroachdb/cockroach/blob/397dc0b48ba959278c0ab9e0042bf5b78940b279/pkg/util/randutil/rand.go#L133

This seed is then what gets printed out as the actual seed being used - which seems misleading, correct?


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114052)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @herkolategan, and @renatolabs)

___
*[`pkg/workload/schemachange/operation_generator.go` line 3321 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/114052#-NqFOYZO1i7H8Nvxa_VT:-NqU7S1ZDtMyLdK7Y8tQ:b47m4rk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ead911a3c37e6f13fa89f74faa20accdbf523/pkg/workload/schemachange/operation_generator.go#L3321)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

good point, but i think this includes extraneous tables that are auto-generated, correct? like table_columns, table_indexes, table_row_statistics, etc.

i can do `LIKE 'table_w%'` perhaps, assuming that no auto-generated table_w* gets added
</blockquote></details>

ah good point. it could be a good place to use SIMILAR TO, which combines LIKE with regex:
```
SELECT 'table_w1_2' SIMILAR TO 'table_w[0-9]_+%';
  ?column?
------------
     t
```
___
*[`pkg/workload/schemachange/schemachange.go` line 199 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/114052#-NqFMW4o0qUt_7phf6f6:-NqU86biEqxUhAbSC9qn:bw4zwzd) ([raw file](https://github.com/cockroachdb/cockroach/blob/84f5446022cc27d2ddb923b850796975d762b61f/pkg/workload/schemachange/schemachange.go#L199)):*
> This seed is then what gets printed out as the actual seed being used - which seems misleading, correct?

i see, nice find.

are you referring to this line? https://github.com/cockroachdb/cockroach/blob/40754196fef9dd70472950f6d493b2a324c261f2/pkg/workload/schemachange/schemachange.go#L205

i agree, that is misleading.

the actual global seed is logged by the the `randutil` package: https://github.com/cockroachdb/cockroach/blob/397dc0b48ba959278c0ab9e0042bf5b78940b279/pkg/util/randutil/rand.go#L250

however, this only gets logged if `randutil.SeedForTests()` is called, and that's only supposed to be called from test code, not workload code.

so let's add a comment above this `NewPseudoRand()` call:
```
// Use NewPseudoRand here because we want to print out the global seed used by the workload.
// Using NewTestRand() here would only let us see the per-test seed that is derived from the global seed.
```


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm after the last nits are fixed!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114052#-:-NqUB2lu9LMEJCnx961E:b-d62w5t)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @herkolategan, and @renatolabs)



<!-- Sent from Reviewable.io -->

annrpom commented:
> are you referring to this line?

yes indeed

also done & done to the changes requested - let's see how this goes 🤞

TFTR! ('-')7
bors r=rafiss
